src/Php/ComposerPhpVersionFactory.php

```diff
@@ -0,0 +1,126 @@
+<?php declare(strict_types = 1);
+
+namespace PHPStan\Php;
+
+use Composer\Semver\VersionParser;
+use Nette\Utils\Json;
+use Nette\Utils\JsonException;
+use Nette\Utils\Strings;
+use PHPStan\File\CouldNotReadFileException;
+use PHPStan\File\FileReader;
+use PHPStan\ShouldNotHappenException;
+use function count;
+use function end;
+use function is_array;
+use function is_file;
+use function is_int;
+use function is_string;
+use function sprintf;
+
+class ComposerPhpVersionFactory
+{
+
+	private ?PhpVersion $minVersion = null;
+
+	private ?PhpVersion $maxVersion = null;
+
+	/**
+	 * @param string[] $composerAutoloaderProjectPaths
+	 * @param int|array{min: int, max: int}|null $phpVersion
+	 */
+	public function __construct(
+		private array $composerAutoloaderProjectPaths,
+		int|array|null $phpVersion,
+		bool $bleedingEdge,
+	)
+	{
+		if (is_int($phpVersion)) {
+			return;
+		}
+
+		if (is_array($phpVersion)) {
+			if ($phpVersion['max'] < $phpVersion['min']) {
+				throw new ShouldNotHappenException('Invalid PHP version range: phpVersion.max should be greater or equal to phpVersion.min.');
```

Comment:
I'd like to validate this sooner. ContainerFactory::validateParameters is the best place to do this.

We can keep this here but when someone puts the wrong value in phpstan.neon, PHPStan should output a nice error message like it does today with invalid parameter values.